File: 2024/day_25/solution.py
def execute(code: str) -> int:
    counter = 0
    pointer = 0

    while pointer < len(code):
        char = code[pointer]

        if char == "+":
            counter += 1
        elif char == "-":
            counter -= 1
        elif char == "[":
            if counter == 0:
                pointer = code.index("]", pointer)
        elif char == "]":
            if counter != 0:
                pointer = code.rindex("[", 0, pointer)
                continue
        elif char == "{":
            if counter == 0:
                pointer = code.index("}", pointer)
        
        pointer += 1

    return counter

File: 2024/day_25/solution.js
function execute(code) {
  let counter = 0;
  let pointer = 0;

  while (pointer < code.length) {
    const char = code[pointer];

    switch (char) {
      case "+":
        counter++;
        break;
      case "-":
        counter--;
        break;
      case "[":
        if (counter === 0) {
          let brackets = 1;
          while (brackets > 0) {
            pointer++;
            if (code[pointer] === "[") brackets++;
            if (code[pointer] === "]") brackets--;
          }
        }
        break;
      case "]":
        if (counter !== 0) {
          let brackets = 1;
          while (brackets > 0) {
            pointer--;
            if (code[pointer] === "]") brackets++;
            if (code[pointer] === "[") brackets--;
          }
          continue;
        }
        break;
      case "{":
        if (counter === 0) {
          let braces = 1;
          while (braces > 0) {
            pointer++;
            if (code[pointer] === "{") braces++;
            if (code[pointer] === "}") braces--;
          }
        }
        break;
    }
    pointer++;
  }

  return counter;
}

console.log(execute("+++")); // 3
console.log(execute("+--")); // -1
console.log(execute(">+++[-]")); // 0
console.log(execute(">>>+{++}")); // 3
console.log(execute("+{[-]+}+")); // 2
console.log(execute("{+}{+}{+}")); // 0
console.log(execute("------[+]++")); // 2
console.log(execute("-[++{-}]+{++++}")); // 5
